### ISSUE TITLE
Firefox crashes when switching between Private browsing and normal browsing tabs

### ISSUE THREAD

**Issue by abodea** [mozilla-mobile/firefox-ios#10738]
## Steps to reproduce
1. Launch Firefox and log in with a valid account.
2. Perform a sync.
3. Tap on the hamburger menu and then access `History`.
4. From any of the sections Today, Yesterday, or Last week, open some websites in Private Browsing (3+ websites).
5. Tap on the `Done` button and access the tab section.
6. In the tab section switch between PB and normal browsing a few times.
### Expected behavior
The opened websites in PB and Normal browsing are displayed without issues and FF doesn't crash.
### Actual behavior
Firefox crashes when switching between Private browsing and normal browsing tabs
### Device & build information
* Device: iPhone 13 Pro Max (15.1.1).
* Build version: Daylight 9000 (9505).
### Notes
Please, note that this issue can be reproduced only when opening websites from history in PB.
Attachments: 

https://user-images.githubusercontent.com/42831109/168109123-f86dae74-4b25-4cd0-bbfa-ad1277dc0e31.mov

```kotlin
{"app_cohort":"2|date=1652346000000&sf=143487&tid=9c4718f7e2b4431f2c25be527c56ff77f9b408eebbb79b18f042fa15435967f6&ttype=i","app_name":"Client","app_version":"9000","timestamp":"2022-05-12 17:38:24.00 +0300","slice_uuid":"840cc738-f5e7-3a3c-a2d9-725a0a7e932c","adam_id":"1073219424","build_version":"9505","platform":2,"bundleID":"org.mozilla.ios.FirefoxBeta","share_with_app_devs":0,"is_first_party":0,"bug_type":"309","os_version":"iPhone OS 15.1.1 (19B81)","incident_id":"45948B18-BA69-49AA-9628-1D04101AD34E","name":"Client","is_beta":1}
{
  "uptime" : 47000,
  "procLaunch" : "2022-05-12 12:15:03.5067 +0300",
  "procRole" : "Foreground",
  "version" : 2,
  "userID" : 501,
  "deployVersion" : 210,
  "modelCode" : "iPhone14,3",
  "procStartAbsTime" : 1036399113241,
  "coalitionID" : 619,
  "osVersion" : {
    "isEmbedded" : true,
    "train" : "iPhone OS 15.1.1",
    "releaseType" : "User",
    "build" : "19B81"
  },
  "captureTime" : "2022-05-12 17:38:23.1031 +0300",
  "incident" : "45948B18-BA69-49AA-9628-1D04101AD34E",
  "bug_type" : "309",
  "pid" : 1823,
  "procExitAbsTime" : 1132188104093,
  "cpuType" : "ARM-64",
  "procName" : "Client",
  "procPath" : "\/private\/var\/containers\/Bundle\/Application\/6C12661F-5B63-4108-AA72-82809DF52E76\/Client.app\/Client",
  "bundleInfo" : {"CFBundleShortVersionString":"9000","CFBundleVersion":"9505","CFBundleIdentifier":"org.mozilla.ios.FirefoxBeta","DTAppStoreToolsBuild":"13F15"},
  "storeInfo" : {"itemID":"1073219424","storeCohortMetadata":"2|date=1652346000000&sf=143487&tid=9c4718f7e2b4431f2c25be527c56ff77f9b408eebbb79b18f042fa15435967f6&ttype=i","entitledBeta":true,"deviceIdentifierForVendor":"022A505A-91E1-4E0E-B426-AB582FA410F2","softwareVersionExternalIdentifier":"88886674","applicationVariant":"1:iPhone14,3:15","thirdParty":true},
  "parentProc" : "launchd",
  "parentPid" : 1,
  "coalitionName" : "org.mozilla.ios.FirefoxBeta",
  "isBeta" : 1,
  "basebandVersion" : "1.15.05",
  "isCorpse" : 1,
  "exception" : {"codes":"0x0000000000000000, 0x0000000000000000","rawCodes":[0,0],"type":"EXC_CRASH","signal":"SIGKILL"},
  "termination" : {"code":2343432205,"flags":6,"namespace":"FRONTBOARD","reasons":["<RBSTerminateContext| domain:10 code:0x8BADF00D explanation:scene-update watchdog transgression: application<org.mozilla.ios.FirefoxBeta>:1823 exhausted real (wall clock) time allowance of 10.00 seconds","ProcessVisibility: Foreground","ProcessState: Running","WatchdogEvent: scene-update","WatchdogVisibility: Foreground","WatchdogCPUStatistics: (","\"Elapsed total CPU time (seconds): 4.890 (user 2.930, system 1.960), 8% CPU\",","\"Elapsed application CPU time (seconds): 0.000, 0% CPU\"",") reportType:CrashLog maxTerminationResistance:Interactive>"]}
  ```

[crashlogs.zip](https://github.com/mozilla-mobile/firefox-ios/files/8680452/crashlogs.zip)

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FXIOS-4231)

**Comment by yoanarios** [comment]
The crash is not related to the history panel is caused to UI work done outside of the main thread when switching between private and normal mode in the tab tray. Tab groups has to be enabled, couldn't reproduce without it

**Comment by badboy** [comment]
Fwiw I seem to run into the same/similar bug, but with even simpler steps to reproduce:

1. Go to the tab overview
2. Click the "Private Browsing" header
3. Click the "Synced Tabs" header
4. Click the "Private Browshing" header
5. click the "Open tabs" header

Boom, it crashes.

**Comment by isabelrios** [comment]
I can reproduce this issue in latest TF build: 9659.
- Open the tab tray
- Go to private mode
- Go back to regular mode 
- App crashes 100%

**Comment by abodea** [comment]
Verified as fixed on the latest version 101.0 (9687) with iPhone 13 Pro max (15.1.1), iPad mini 5 (14.7.1).
Note that I tried my scenario from the description but also from the comments above.

**Comment by badboy** [comment]
Any chance we can get that build onto TestFlight?

**Comment by lmarceau** [comment]
@badboy this fix is currently in Nightly build (v9000)

**Comment by badboy** [comment]
I‘m on the beta TestFlight (v101, 9619).
Maybe I should switch.

**Comment by SimonBasca** [comment]
@badboy this was fixed on version 101.0 (9687). Could you please try upgrading to  101.0 (9687) the latest 101.0 (9723)

**Comment by badboy** [comment]
The latest version I see in my TestFlight right now is 101.0 (9619).
